### PR TITLE
add test_default_route_with_bgp_flap

### PR DESCRIPTION
**Pull request by arlakshm**
Signed-off-by: Arvindsrinivasan Lakshmi Narasimhan <arlakshm@microsoft.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change
Add a new test to verify the fix for https://github.com/Azure/sonic-buildimage/issues/8979 and https://github.com/Azure/sonic-buildimage/issues/9055
<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ x] Test case(new/improvement)


### Back port request
- [ ] 201911

### Approach
#### What is the motivation for this PR?
This PR adds a new test to check the default route is correct in APP_DB

#### How did you do it?
The test check the following
 - get the default route from app_db and check if the nexthop ips match that of ip interfaces of the peer mentioned in the testbed
 - Shutdown all the bgp sessions and verify the default route is removed from APP_DB
 - 
#### How did you verify/test it?
Test on vs:
```
arlakshm@7d00946cb266:/data/sonic_mgmt_int/tests$ ./run_tests.sh -n vms-kvm-t0 -d vlab-01 -c route/test_default_route.py::test_default_route_with_bgp_flap[vlab-01-None]  -f vtestbed.csv -i veos_vtb -u -e "--skip_sanity --disable_loganalyzer" -l info
=== Running tests in groups ===
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
=============================================================================================================================== test session starts ================================================================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /data/sonic_mgmt_int/tests, inifile: pytest.ini
plugins: metadata-1.11.0, repeat-0.9.1, forked-1.3.0, xdist-1.28.0, html-1.22.1, ansible-2.2.2
collecting ...
------------------------------------------------------------------------------------------------------------------------------- live log collection --------------------------------------------------------------------------------------------------------------------------------
18:49:45 __init__.load_dut_basic_facts            L0080 INFO   | Getting dut basic facts
18:49:54 facts_cache.read                         L0093 INFO   | Load cache file "/data/sonic_mgmt_int/tests/_cache/vms-kvm-t0/tbinfo.pickle" failed with exception: IOError(2, 'No such file or directory')
18:49:55 facts_cache.write                        L0114 INFO   | Create cache dir /data/sonic_mgmt_int/tests/_cache/vms-kvm-t0
18:49:55 facts_cache.write                        L0120 INFO   | Cached facts "vms-kvm-t0.tbinfo" to /data/sonic_mgmt_int/tests/_cache/vms-kvm-t0/tbinfo.pickle
18:49:55 facts_cache.read                         L0093 INFO   | Load cache file "/data/sonic_mgmt_int/tests/_cache/vlab-01/host_visible_vars.pickle" failed with exception: IOError(2, 'No such file or directory')
18:49:55 facts_cache.write                        L0114 INFO   | Create cache dir /data/sonic_mgmt_int/tests/_cache/vlab-01
18:49:55 facts_cache.write                        L0120 INFO   | Cached facts "vlab-01.host_visible_vars" to /data/sonic_mgmt_int/tests/_cache/vlab-01/host_visible_vars.pickle
18:49:55 conftest.generate_param_asic_index       L0856 INFO   | generating num_asics asic indicies for  DUT [['vlab-01']] in
18:49:55 conftest.generate_param_asic_index       L0871 INFO   | dut name vlab-01  asics params = [None]
18:49:55 conftest.generate_param_asic_index       L0856 INFO   | generating num_asics asic indicies for  DUT [['vlab-01']] in
18:49:55 conftest.generate_param_asic_index       L0871 INFO   | dut name vlab-01  asics params = [None]
18:49:55 conftest.generate_param_asic_index       L0856 INFO   | generating num_asics asic indicies for  DUT [['vlab-01']] in
18:49:55 conftest.generate_param_asic_index       L0871 INFO   | dut name vlab-01  asics params = [None]
collecting 1 item                                                                                                                                                                                                                                                                  18:49:55 __init__.pytest_collection_modifyitems   L0279 INFO   | Available basic facts that can be used in conditional skip:
{
  "commit_id": "36d866002",
  "build_date": "Sun Dec 26 18:33:50 UTC 2021",
  "topo_type": "t0",
  "hwsku": "Force10-S6000",
  "build_number": 61557,
  "asic_type": "vs",
  "sonic_utilities": 1.2,
  "kernel_version": "5.10.0-8-2-amd64",
  "debian_version": "11.2",
  "built_by": "AzDevOps@sonic-build-workers-0010KF",
  "num_asic": 1,
  "is_multi_asic": false,
  "platform": "x86_64-kvm_x86_64-r0",
  "branch": "master",
  "libswsscommon": "1.0.0",
  "build_version": "master.61557-dirty-20211226.182606",
  "release": "master",
  "topo_name": "t0"
}
18:49:55 __init__.pytest_collection_modifyitems   L0282 INFO   | Processing: route/test_default_route.py::test_default_route_with_bgp_flap[vlab-01-None]

---------------------------------------------------------------------------------------------------------------------------------- live log setup ----------------------------------------------------------------------------------------------------------------------------------
18:49:55 __init__.set_default                     L0049 INFO   | Completeness level not set during test execution. Setting to default level: CompletenessLevel.basic
18:49:55 __init__.check_test_completeness         L0139 INFO   | Test has no defined levels. Continue without test completeness checks
18:49:55 facts_cache.read                         L0093 INFO   | Load cache file "/data/sonic_mgmt_int/tests/_cache/vlab-01/basic_facts.pickle" failed with exception: IOError(2, 'No such file or directory')
18:50:52 facts_cache.write                        L0120 INFO   | Cached facts "vlab-01.basic_facts" to /data/sonic_mgmt_int/tests/_cache/vlab-01/basic_facts.pickle
18:51:16 conftest.generate_params_dut_hostname    L0891 INFO   | Using DUTs ['vlab-01'] in testbed 'vms-kvm-t0'
18:51:16 conftest.rand_one_dut_hostname           L0301 INFO   | Randomly select dut vlab-01 for testing
18:51:16 conftest.creds_on_dut                    L0527 INFO   | dut vlab-01 belongs to groups [u'lab', u'sonic', 'fanout']
18:51:16 conftest.creds_on_dut                    L0539 INFO   | skip empty var file ../ansible/group_vars/all/corefile_uploader.yml
18:51:16 conftest.creds_on_dut                    L0539 INFO   | skip empty var file ../ansible/group_vars/all/env.yml
18:51:20 __init__.sanity_check                    L0130 INFO   | Prepare sanity check
18:51:20 __init__.sanity_check                    L0140 INFO   | Found marker: m.name=topology, m.args=('any',), m.kwargs={}
18:51:20 __init__.sanity_check                    L0140 INFO   | Found marker: m.name=device_type, m.args=('vs',), m.kwargs={}
18:51:20 __init__.sanity_check                    L0166 INFO   | Skip sanity check according to command line argument or configuration of test script.
18:51:20 __init__._fixture_func_decorator         L0058 INFO   | -------------------- fixture skip_test_module_over_backend_topologies setup starts --------------------
18:51:20 __init__._fixture_func_decorator         L0065 INFO   | -------------------- fixture skip_test_module_over_backend_topologies setup ends --------------------
18:51:20 __init__.loganalyzer                     L0047 INFO   | Log analyzer is disabled
---------------------------------------------------------------------------------------------------------------------------------- live log call -----------------------------------------------------------------------------------------------------------------------------------
18:51:29 test_default_route.verify_default_route_ L0079 INFO   | default route from app db {u'ROUTE_TABLE:0.0.0.0/0': {u'type': u'hash', u'value': {u'ifname': u'PortChannel0001,PortChannel0002,PortChannel0003,PortChannel0004', u'nexthop': u'10.0.0.57,10.0.0.59,10.0.0.61,10.0.0.63'}, u'expireat': 1640717483.3213634, u'ttl': -0.001}}
18:51:29 test_default_route.verify_default_route_ L0085 INFO   | nexthop list in app_db 10.0.0.57,10.0.0.59,10.0.0.61,10.0.0.63
18:51:29 test_default_route.get_upstream_neigh    L0040 INFO   | testbed topo t0 upstream neigh type t1
18:51:29 test_default_route.verify_default_route_ L0095 INFO   | peer intf ip from tb set(['10.0.0.59', '10.0.0.61', '10.0.0.63', '10.0.0.57'])
18:51:34 test_default_route.verify_default_route_ L0079 INFO   | default route from app db {u'ROUTE_TABLE:::/0': {u'type': u'hash', u'value': {u'ifname': u'PortChannel0001,PortChannel0002,PortChannel0003,PortChannel0004', u'nexthop': u'fc00::72,fc00::76,fc00::7a,fc00::7e'}, u'expireat': 1640717489.3764486, u'ttl': -0.001}}
18:51:34 test_default_route.verify_default_route_ L0085 INFO   | nexthop list in app_db fc00::72,fc00::76,fc00::7a,fc00::7e
18:51:34 test_default_route.get_upstream_neigh    L0040 INFO   | testbed topo t0 upstream neigh type t1
18:51:34 test_default_route.verify_default_route_ L0095 INFO   | peer intf ip from tb set(['fc00::7e', 'fc00::76', 'fc00::7a', 'fc00::72'])
18:54:27 config_reload.config_reload              L0058 INFO   | reloading config_db
PASSED                                                                                                                                                                                                                                                                       [100%]

------------------------------------------------------------------------------------------------------------ generated xml file: /data/sonic_mgmt_int/tests/logs/tr.xml ------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------ live log sessionfinish ------------------------------------------------------------------------------------------------------------------------------
18:59:21 __init__.pytest_terminal_summary         L0064 INFO   | Can not get Allure report URL. Please check logs
========================== 1 passed in 575.60 seconds ==========================
INFO:root:Can not get Allure report URL. Please check logs
arlakshm@7d00946cb266:/data/sonic_mgmt_int/tests$
```

#### Any platform specific information?
No
#### Supported testbed topology if it's a new test case?
Any
### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

### CODE REVIEW

**Comment by arlakshm** [comment]
@abdosi, @wangxin, @prsunny can you please review this PR and provide feedback.

**Review by abdosi — DISMISSED** [review]
LGTM

File: tests/route/test_default_route.py [hunk 1]
```diff
@@ -62,3 +145,38 @@ def test_default_ipv6_route_next_hop_global_address(duthosts, enum_rand_one_per_
...
+    for af in af_list:
+        verify_default_route_in_app_db(asichost, tbinfo, af)
+
+    duthost.command("sudo config bgp shutdown all")
+    if not wait_until(120, 2, 0, duthost.is_bgp_state_idle):
+        pytest.fail(
+            'BGP Shutdown Timeout: BGP sessions not shutdown after 120 seconds')
+
+    # give some more time for default route to be removed
+    time.sleep(120)
```
Comment (lolyu) [inline tests/route/test_default_route.py:175]: how about replacing this sleep with `wait_until` to constantly check the default route removal?

File: tests/route/test_default_route.py [hunk 2]
```diff
@@ -62,3 +145,38 @@ def test_default_ipv6_route_next_hop_global_address(duthosts, enum_rand_one_per_
...
+    try:
+        for af in af_list:
+            default_route = asichost.get_def_route_frm_app_db(af)
+            pytest_assert(not default_route, \
+                "default route present after bgp shutdown: {}".format(default_route))
+    finally:
+        config_reload(duthost)
```
Comment (lolyu) [inline tests/route/test_default_route.py:182]: needs an empty line here

File: tests/common/devices/sonic_asic.py [hunk 3]
```diff
@@ -560,3 +561,17 @@ def get_bgp_statistic(self, stat):
...
+            def_rt_str = 'ROUTE_TABLE:0.0.0.0/0'
+        else:
+            def_rt_str = 'ROUTE_TABLE:::/0'
+
+        def_rt_entry = self.sonichost.shell(
+            "{} redis-dump -y -k \"{}\" --pretty".format(
+                self.ns_arg, def_rt_str))['stdout']
+        if def_rt_entry is not None:
+            def_rt_json = json.loads(def_rt_entry)
+        return def_rt_json
```
Comment (lolyu) [inline tests/common/devices/sonic_asic.py:577]: empty line here?

File: tests/common/devices/sonic_asic.py [hunk 4]
```diff
@@ -560,3 +561,17 @@ def get_bgp_statistic(self, stat):
     def check_bgp_statistic(self, stat, value):
         val = self.get_bgp_statistic(stat)
         return val == value
+
+    def get_def_route_frm_app_db(self, af='ipv4'):
```
Comment (lolyu) [inline tests/common/devices/sonic_asic.py:565]: suggest not using abbreviation, like `get_default_route_from_app_db`

File: tests/route/test_default_route.py (hunk 2 above)
Comment (arlakshm) [reply tests/route/test_default_route.py]: added in latest commit

File: tests/common/devices/sonic_asic.py (hunk 3 above)
Comment (arlakshm) [reply tests/common/devices/sonic_asic.py]: added in latest commit

File: tests/common/devices/sonic_asic.py (hunk 4 above)
Comment (arlakshm) [reply tests/common/devices/sonic_asic.py]: fixed in latest commit

File: tests/route/test_default_route.py (hunk 1 above)
Comment (arlakshm) [reply tests/route/test_default_route.py]: address in latest commit

**Review by lolyu — APPROVED** [review]

File: tests/route/test_default_route.py [hunk 5]
```diff
@@ -13,7 +16,88 @@
...
+    upstream_neigh = get_upstream_neigh(tbinfo)
+    pytest_assert(upstream_neigh is not None, "No upstream neighbors in the testbed")
+
+    if af == 'ipv4':
+        upstream_neigh_ip = set([upstream_neigh[neigh][0]  for neigh in upstream_neigh])
+    elif af == 'ipv6':
+        upstream_neigh_ip= set([upstream_neigh[neigh][1]  for neigh in upstream_neigh])
+
+    logging.info("peer intf ip from tb {}".format(upstream_neigh_ip))
+    pytest_assert(len(nexthops) == len(upstream_neigh_ip), \
```
Comment (wenyiz2021) [inline tests/route/test_default_route.py:94]: Hi Arvind, this won't work for multi-asic if there are 2 backend asics that both connect to upstream neighbors.
While nexthops are coming from one asic only.

**Pull request merged**